tools/pd-analysis/transfer_region_counter.go

```diff
@@ -0,0 +1,178 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package analysis
+
+import (
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+)
+
+// TransferRegionCount is to count transfer schedule for judging whether redundant
+type TransferRegionCount struct {
+	storeNum        int
+	regionNum       int
+	IsValid         bool
+	Redundant       uint64
+	Necessary       uint64
+	regionMap       map[uint64]uint64
+	visited         []bool
+	GraphMat        [][]uint64
+	mutex           sync.Mutex
+	loopResultPath  [][]int
+	loopResultCount []uint64
+}
+
+// TransferRegionCounter is an global instance for TransferRegionCount
+var TransferRegionCounter TransferRegionCount
+
+// Init for TransferRegionCount
+func (c *TransferRegionCount) Init(storeNum, regionNum int) {
+	c.storeNum = storeNum
+	c.regionNum = regionNum
+	c.IsValid = true
+	c.Redundant = 0
+	c.Necessary = 0
+	c.regionMap = make(map[uint64]uint64)
+	c.visited = make([]bool, c.storeNum+1)
+	for i := 0; i < c.storeNum+1; i++ {
+		tmp := make([]uint64, c.storeNum+1)
+		c.GraphMat = append(c.GraphMat, tmp)
+	}
+	c.loopResultPath = c.loopResultPath[:0]
+	c.loopResultCount = c.loopResultCount[:0]
+}
+
+// AddTarget is be used to add target of edge in graph mat.
+// Firstly add a new peer and then delete the old peer of the scheduling,
+// So in the statistics, also firstly add the target and then add the source.
+func (c *TransferRegionCount) AddTarget(regionID, targetStoreID uint64) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	c.regionMap[regionID] = targetStoreID
+}
+
+// AddSource is be used to add source of edge in graph mat.
+func (c *TransferRegionCount) AddSource(regionID, sourceStoreID uint64) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	if targetStoreID, ok := c.regionMap[regionID]; ok {
+		c.GraphMat[sourceStoreID][targetStoreID]++
+		delete(c.regionMap, regionID)
+	} else {
+		fmt.Println("Error when add sourceStoreID %u with regionID %u in transfer region map", sourceStoreID, regionID)
+		os.Exit(-1)
+	}
+}
+
+//DFS is used to find all the looped flow in such a directed graph.
+//For each point U in the graph, a DFS is performed, and push the passing point v
+//to the stack. If there is an edge of `v->u`, then the corresponding looped flow
+//is marked and removed. When all the output edges of the point v are traversed,
+//pop the point v out of the stack.
+func (c *TransferRegionCount) DFS(cur int, curFlow uint64, path []int) {
+	//push stack
+	path = append(path, cur)
+	c.visited[cur] = true
+
+	for target := path[0]; target < c.storeNum+1; target++ {
```

Comment:
Actually, the TiDB cluster store number do not same as the simulator, the store_id may likes [1, 4, 10230, 10232 .....]

Reply:
Thanks, I have added support for sparse stores